### PR TITLE
Shengwei_HotFix: Page render issue in Timelog

### PR DESCRIPTION
**Pull request by ptrpengdev**
# Description
Hotfix blank page in the main site due to the user's timelog entry is linked to a project from which the user has been removed.

image

Fix rendering issue in Timelog
- Ensure that the page renders all time entries regradless of projects the user is assigned to
- Add a warning to time entries associated with projects that the user no longer has access to
- Prevent time entries from being updated if associated with projects that the user no longer has access to.

## Related PRS (if any):
Backend [PR898](https://github.com/OneCommunityGlobal/HGNRest/pull/898)

## Main changes explained:
- Update file src/components/Timelog/TimeEntry.jsx for including new logic for time entries rendering.
hotfix_main_time_entry_issue

## How to test:
1. check into current branch and checkout backend [PR898](https://github.com/OneCommunityGlobal/HGNRest/pull/898)
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log as admin user
5. go to dashboard -> timelog
6. log a time entry under a specific project
7. remove the user from the project 
8. reload the timelog page

## Note:
Include the information the reviewers need to know.

### CODE REVIEW

File: src/components/Timelog/TimeEntry.jsx [hunk 1]
```diff
@@ -161,7 +170,7 @@ const TimeEntry = (props) => {
             </p>
             <div className='mb-3'>
             {
-              canEdit 
+              (canEdit && (from === 'WeeklyTab' ? !shouldPreventTimeEntryUpdateInWeeklyTab : true))
```
Comment (mSharifHub) [inline src/components/Timelog/TimeEntry.jsx:173]: const canEdit = (dispatch(hasPermission('editTimelogInfo')) 
    //permission to edit any time entry on their own time logs tab
    || dispatch(hasPermission('editTimeEntry'))) && !cantEditJaeRelatedRecord. 
    
    Every call to dispatch can potentially trigger a redux state update and cause react to re-render the component. Dispatching actions in the render phase can lead to unexpected or duplicate side effects. 

```suggestion
   const dispatch = useDispatch();
  const [canEdit, setCanEdit] = useState(false);
  const [canDelete, setCanDelete] = useState(false);
  
  useEffect(() => {
    const checkPermissions = async () => {
      const editPermission = await dispatch(hasPermission('editTimelogInfo')) || await dispatch(hasPermission('editTimeEntry'));
      const deletePermission = await dispatch(hasPermission('deleteTimeEntryOthers')) || (props.isAuthUser && await dispatch(hasPermission('deleteTimeEntry')));
      setCanEdit(editPermission && !props.cantEditJaeRelatedRecord);
      setCanDelete(deletePermission && !props.cantEditJaeRelatedRecord && props.isAuthUserAndSameDayEntry);
    };

    checkPermissions();
  }, [dispatch, props.isAuthUser, props.cantEditJaeRelatedRecord, props.isAuthUserAndSameDayEntry]);

```

**Review by mSharifHub — CHANGES_REQUESTED** [review]
changes may need if is causing application to re-render when  I tested locally and have not came across unnecessary re-renders. however,  initiate any dispatch that check permission in this case in the useEffect hook.  This could possible prevent any re-rendering of page.

**Review by cgomezhub — APPROVED** [review]
Good job! I  the requested changes are completed,

Ensure that the page renders all time entries regradless of projects the user is assigned to

  - good


Add a warning to time entries associated with projects that the user no longer has access to

    - well done



Prevent time entries from being updated if associated with projects that the user no longer has access to.

     - well done
Screenshot 2024-05-01 120359
Screenshot 2024-05-01 120453

File: src/components/Timelog/TimeEntry.jsx (hunk 1 above)
Comment (ptrpengdev) [reply src/components/Timelog/TimeEntry.jsx:173]: The hasPermission function only retrieves data from the Redux store, comparing given permissions with a list of permissions without modifying any state in Redux. 

Regarding the original approach where canDelete/canEdit is declared as a const variable within the component body,  these variables are calculated only once during the initial rendering of the component. Subsequent re-renders won't trigger recalculation unless the component itself is re-rendered due to changes in its parent component's state or props. In such a case, I don't think that using the original approach could lead to unexpected or duplicate side effects.

Perhaps, considering the nature of the hasPermission function — only accessing data from the Redux store without dispatching any actions, maybe we could refactor the hasPermission function into a pure function.

```
const hasPermission = (action) => {
  return (dispatch, getState) => {
    const state = getState();
    const rolePermissions = state.role.roles;
    const userRole = state.auth.user.role;
    const userPermissions = state.auth.user.permissions?.frontPermissions;

    if (userRole && rolePermissions && rolePermissions.length != 0) {
      const roleIndex = rolePermissions?.findIndex(({ roleName }) => roleName === userRole);
      let permissions = [];
      if (roleIndex !== -1) {
        permissions = rolePermissions[roleIndex].permissions;
      }

      return userPermissions?.includes(action) || permissions?.includes(action);
    }
    return false;
  }
};
```

into 

```
const component = (props) => {
  const hasPermission = useMemo(() => {
    return (permissionName) => {
      const state = props.getState();
      const rolePermissions = state.role.roles;
      const userRole = state.auth.user.role;
      const userPermissions = state.auth.user.permissions?.frontPermissions;

      if (userRole && rolePermissions && rolePermissions.length !== 0) {
        const roleIndex = rolePermissions.findIndex(({ roleName }) => roleName === userRole);
        let permissions = [];
        if (roleIndex !== -1) {
          permissions = rolePermissions[roleIndex].permissions;
        }

        return userPermissions?.includes(permissionName) || permissions?.includes(permissionName);
      }
      return false;
    };
  }, [props.getState]);

```
However, the hasPermission function is widely-used throughout the application, and its modification falls outside the scope of this PR. Please correct me if I am wrong. I look forward to hearing your thoughts on this.

**Comment by KaushikMreddy** [comment]
Hi, is this PR to be reviewed or not reviewed? The backend and front-end PRs have separate views being shared.

**Pull request closed without merging**